internal/xdscache/v3/snapshot.go

```diff
@@ -15,60 +15,106 @@ package v3
 
 import (
 	"context"
-	"reflect"
 
+	envoy_service_discovery_v3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
 	envoy_types "github.com/envoyproxy/go-control-plane/pkg/cache/types"
 	envoy_cache_v3 "github.com/envoyproxy/go-control-plane/pkg/cache/v3"
 	envoy_resource_v3 "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
 	"github.com/google/uuid"
 	"github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
 
 	"github.com/projectcontour/contour/internal/dag"
 	contour_xds_v3 "github.com/projectcontour/contour/internal/xds/v3"
 	"github.com/projectcontour/contour/internal/xdscache"
 )
 
 // SnapshotHandler responds to DAG builds via the OnChange()
-// event and generates and caches go-control-plane Snapshots.
+// event and Endpoint updates via the Refresh() event and
+// generates and caches go-control-plane Snapshots.
 type SnapshotHandler struct {
-	// SnapshotCache contains go-control-plane Snapshots
-	// and is used by the go-control-plane xDS server.
-	SnapshotCache envoy_cache_v3.SnapshotCache
-
-	// resources contains the Contour xDS resource caches.
-	resources map[envoy_resource_v3.Type]xdscache.ResourceCache
-	log       logrus.FieldLogger
+	resources    map[envoy_resource_v3.Type]xdscache.ResourceCache
+	defaultCache envoy_cache_v3.SnapshotCache
+	edsCache     envoy_cache_v3.SnapshotCache
```

Comment:
I switched to a SnapshotCache here instead of a LinearCache for EDS. It turns out the LinearCache doesn't behave as desired for us on control plane restarts. If Envoy already has config for a given resource at a particular version, then on a control plane restart, the version number of the resource will be reset to 1 (or close to 1), therefore will not be sent to Envoy since Envoy already has a "later" version of the resource. It will only be sent once the resource changes. However, this means that any changes that happened during the time of the restart may be ignored for a period of time. 

With a SnapshotCache, any time the cached version differs from the last version Envoy saw, the resource will be re-sent, which better fits our use case given that versions are not durable across control plane restarts.